### PR TITLE
Numa optimization and pinning fixes

### PR DESCRIPTION
**Pull request by krenzland**
- Pinning now works even when using more than 1 mpi rank/node
- Initialize DR in a NUMA-aware manner

Improves performance by a lot for R&S-FL when using 1 rank per node on NG.

(History is pretty messy)

### CODE REVIEW

File: src/Parallel/Pin.cpp [hunk 1]
```diff
@@ -65,28 +81,19 @@ cpu_set_t seissol::parallel::getWorkerUnionMask() {
...
-  cpu_set_t set;
-  CPU_ZERO(&set);
-  for (int i = 0; i < get_nprocs() ; ++i) {
-    CPU_SET(i, &set);
-  }
-  CPU_XOR(&set, &set, &workerUnion);
-
-  return set;
+cpu_set_t seissol::parallel::Pinning::getFreeCPUsMask() const {
+  return pthreadMask;
```
Comment (uphoffc) [inline src/Parallel/Pin.cpp:85]: (You can ignore the following comment if you want.)

Storing pthreadMask is actually redundant as you already store processMask and openmpMask. The operation to compute pthreadMask is really cheap.

File: src/Parallel/Pin.h [hunk 2]
```diff
@@ -45,11 +45,23 @@
...
+private:
+  cpu_set_t processMask{};
+  cpu_set_t pthreadMask{};
+  cpu_set_t openmpMask;
+public:
+  Pinning();
+  void init();
+
+  cpu_set_t getProcessMask() const { return processMask; };
+  cpu_set_t getPthreadMask() const { return pthreadMask; };
```
Comment (uphoffc) [inline src/Parallel/Pin.h:58]: getPthreadMask and getFreeCPUsMask are aliases?

File: src/SeisSol.h [hunk 3]
```diff
@@ -134,7 +140,11 @@ class SeisSol
 	}
 
 public:
-	/**
+  parallel::Pinning getPinning() {
+	  return pinning;
+	}
```
Comment (uphoffc) [inline src/SeisSol.h:145]: Do you intend to return a copy here? Should be ok as the object is not large.

File: src/Solver/time_stepping/TimeManager.cpp [hunk 4]
```diff
@@ -409,7 +410,7 @@ void seissol::time_stepping::TimeManager::pollForCommunication() {
   // pin this thread to the last core
   volatile unsigned int l_signalSum = 0;
 
-  parallel::pinToFreeCPUs();
+  seissol::SeisSol::main.getPinning().pinToFreeCPUs();
```
Comment (uphoffc) [inline src/Solver/time_stepping/TimeManager.cpp:413]: Same comment as above, I'd pass pinning as parameter.

**Review by uphoffc — CHANGES_REQUESTED** [review]

File: src/Physics/ini_model_DR.f90 [hunk 5]
```diff
@@ -189,13 +213,13 @@ SUBROUTINE DR_basic_ini(DISC,EQN,MESH,BND)                 ! global variables
...
      ELSEIF ((DISC%DynRup%RFtime_on .EQ. 1) .AND. (DISC%DynRup%DS_output_on .EQ. 0 )) THEN
-           DO i = 1, MESH%Fault%nSide
+        DO i = 1, MESH%Fault%nSide
               IF (MESH%FAULT%Face(i,1,1) .NE. 0) THEN
                  DISC%DynRup%RF(:,i) = .TRUE.
               ENDIF
           ENDDO
     ENDIF
-    if (EQN%FL .NE. 33) then !33 is ImposedSlipRateOnDRBoundary
+    if (EQN%FL /= 33) then !33 is ImposedSlipRateOnDRBoundary
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Physics/ini_model_DR.f90:222]: No, apparently .NE. is deprecated

File: src/SeisSol.cpp [hunk 6]
```diff
@@ -60,15 +60,28 @@
...
 #include "version.h"
 
+#ifdef USE_ASAGI
+#include "Reader/AsagiModule.h"
+#endif
+
 bool seissol::SeisSol::init(int argc, char* argv[])
 {
+  // Initialize pinning. Needs to be called before the first OpenMP rank!
+  // (Doesn't actually pin anything, yet!)
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/SeisSol.cpp:70]: True.

File: src/Initializer/dg_setup.f90 [hunk 7]
```diff
@@ -535,6 +540,20 @@ SUBROUTINE iniGalerkin3D_us_level2_new(OptionalFields,EQN,DISC,MESH,BND,IC,SOURC
       allocate(disc%DynRup%output_dynStress_time(DISC%Galerkin%nBndGP,MESH%Fault%nSide))      
       allocate(disc%DynRup%output_StateVar(DISC%Galerkin%nBndGP,MESH%Fault%nSide))
 
+      ! Initialize w/ first-touch
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Initializer/dg_setup.f90:543]: Not sure. The initialization is all over the code...

File: src/Checkpoint/Manager.cpp [hunk 8]
```diff
@@ -134,4 +134,17 @@ bool seissol::checkpoint::Manager::init(real* dofs, unsigned int numDofs,
 		removeBuffer(FILENAME);
 
 		return exists;
+}
+
+void seissol::checkpoint::Manager::setUp()
+{
+  setExecutor(m_executor);
+  if (isAffinityNecessary()) {
+    const auto freeCpus = SeisSol::main.getPinning().getFreeCPUsMask();
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Checkpoint/Manager.cpp:143]: Sure, if you can tell me where this stuff is actually called. If I remember correctly, it's quite hidden.

[the remark this answers is not in the document]
Comment (uphoffc) [reply src/Checkpoint/Manager.cpp:143]: It's called by async. If you want to change it you'd need to pass pinning to the constructor. Might make sense to introduce a base class here, as the pinning code is copied & pasted as far as I could see.

File: src/Physics/ini_model_DR.f90 [hunk 9]
```diff
@@ -163,15 +163,39 @@ SUBROUTINE DR_basic_ini(DISC,EQN,MESH,BND)                 ! global variables
...
+
+    IF (DISC%DynRup%magnitude_output_on == 1) THEN
        ALLOCATE(  DISC%DynRup%averaged_Slip(MESH%Fault%nSide)        )
        !ini magnitude output
-       DISC%DynRup%magnitude_out(:) = .TRUE.
-       DISC%DynRup%averaged_Slip(:) = 0.0D0
+
+       !$omp parallel do schedule(static)
+       DO i=1,MESH%fault%nSide
+           DISC%DynRup%magnitude_out(i) = .TRUE.
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Physics/ini_model_DR.f90:196]: I really don't want to know.

**Review by uphoffc — APPROVED** [review]

**Pull request merged**